Fixes #32678 - katello_ca_consumer in registration template

stejskalleos opened the pull request:
Move `rhsm_reconfigure` script [0] from `katello_consumer.rpm` to
`global_registration` template so the `rpm` is not needed anymore

Migrated script is without support of RHEL5 and older
`subscription-manager` versions (0.96 and bellow)

[0] [rhsm_reconfigure_script](https://github.com/theforeman/puppet-foreman_proxy_content/blob/master/lib/puppet/provider/rhsm_reconfigure_script/rhsm_reconfigure_script.rb#L32)

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

stejskalleos committed:
Fixes #32678 - katello_ca_consumer in registration template

Move `rhsm_reconfigure` script from `katello_consumer.rpm` to
`global_registration` template so the `rpm` is not needed anymore

Migrated script is without support of RHEL5 and older
`subscription-manager` versions (0.96 and bellow)

ekohl commented on app/controllers/concerns/foreman/controller/registration.rb:
```diff
@@ -82,19 +81,23 @@ def not_found(options = nil)
 
+  def context_urls
+    { url: url,
+      url_host: url.host,
+      url_port: url.port,
```
Isn't this redundant? You already pass the `url` object. If you can't use `url.host` in the template, then passing `url` feels useless.

I do recall the last time I played with the URL object we don't have a proper jail for it. Would you mind looking into that instead?

stejskalleos replied:
As you say, we don't have a jail for URL object so that's why I went this way, plus
wasn't sure if community wants to have jail for URL in the template.

I guess it could be useful, I'll add it.

ekohl commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -114,26 +114,51 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
+    --server.port="<%= @url_port %>" \
+    --server.prefix="/rhsm" \
+    --rhsm.repo_ca_cert="$KATELLO_SERVER_CA_CERT" \
+    --rhsm.baseurl="<%= @url %>pulp/content/"
```
I don't think this is right. This looks like a Pulp 2 URL.

I think there should be a macro from Katello for the correct content URL (note that I'm not saying there is one today, but there should be).

stejskalleos replied:
From this PRs [0][1] & redmine [2] looks like all content can be get from `/pulp/content`

[0] https://github.com/Katello/katello/pull/9261
[1] https://github.com/theforeman/puppet-foreman_proxy_content/commit/28819d0ae1cfea204a7fe2c80093d88d0f2918e2
[2] https://projects.theforeman.org/issues/31867

ekohl replied:
Ah you're right. However, I still thing a macro is needed. My reasoning is that it should respect the Pulp content URL that Katello can discover via the Smart Proxy.

Today it's not supported, but it is my desire to support complex architectures where a Smart Proxy is on a different host than the Pulp instance. There is currently a working group looking at those advanced scenarios. That why I think a `pulp_content_url(host)` makes sense.

While we're at it, a `rhsm_url(host)` macro is also better than assuming it's on the same URL as the Smart Proxy.

And yes, I do believe Katello may be doing the same thing wrong right now. That's why I won't block on it, but I'm working on this on all fronts.

ekohl commented on todo_subman_community.sh:
```diff
@@ -0,0 +1,9 @@
+Subman on debian & ubuntu
```
Did you intend to add this file?

stejskalleos replied:
Nope, mistake :/

ekohl replied:
What I tend to do is type `git status` a lot. Also, `git show` after I've committed to review the diff.

stejskalleos replied:
The sad truth is that I did exactly what you said and still somehow missed it :/

ekohl commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -114,26 +114,51 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
+  fi
+
+  # Configure subscription-manager
+  test -f $CFG.BAK || cp $CFG $CFG.BAK
```
I think we commonly use lower case extensions
```suggestion
  test -f $CFG.bak || cp $CFG $CFG.bak
```

ekohl commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -114,26 +114,51 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
+  <% end -%>
+
+  KATELLO_SERVER_CA_CERT=/etc/rhsm/ca/<%= @url_host %>.pem
+  CFG=/etc/rhsm/rhsm.conf
```
I always like to be a bit more explicit in naming
```suggestion
  RHSM_CFG=/etc/rhsm/rhsm.conf
```

ekohl commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -114,26 +114,51 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
 
 }
```
Unrelated, but this closing bracket's incorrect indenting is very confusing.

ekohl commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -114,26 +114,51 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
-
-    subscription-manager register <%= '--force' if @force %> --org='<%= @organization.label %>' --activationkey='<%= activation_keys %>' || <%= @ignore_subman_errors ? 'true' : 'exit 1' %>
-    register_katello_host | bash
+  <% if @force -%>
```
The indenting here looks odd. Everywhere else 4 spaces are used.

ehelms commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -112,25 +112,50 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
+      --rhsm.repo_ca_cert="$KATELLO_SERVER_CA_CERT" \
+      --rhsm.baseurl="<%= @url %>pulp/content"
+
+    if grep --quiet full_refresh_on_yum $RHSM_CFG; then
```
I think it would be worth a note here saying that we should investigate dropping this block once EL6 support is dropped.

ehelms commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -112,25 +112,50 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
 
     <% if @force -%>
-    yum remove -y katello-ca-consumer*
+    yum remove -y katello-ca-consumer* subscription-manager
```
I think with this change you can drop this whole force block.

ehelms replied:
Additionally, if this template is what configured subscription-manager, I lean towards removing any installed consumer RPMs off the system so as not to give a false impression that they are affecting the client.

stejskalleos replied:
This is only for case when user run registration with `force=true`, which he/she knows that deletes these files. It was also requested by support in BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1921055

ehelms replied:
I am leaning towards we should be removing `katello-ca-consumer*` regardless of whether a parameter is provided. This is about registration, so we should assume that said system is *not* registered or configured correctly at the time the user wants to register the system and cleanup any potential issues on the system.

The idea of a `force` parameter is a holdover concept from subscription-manager where if you were on the client and wanted to either change the smart-proxy it was registered to or change the Content View / Lifecycle environment you had to use the `--force` parameter to basically "recreate" the system. I think we should evolve to look at registration as bringing a machine previously unknown to Foreman to Foreman's inventory and thus cleanup any potential old stuff we know might exist and cause problems and lay down proper configuration based on the users parameters. For changes that happen after registration, that's what other stories and mechanisms should be used for (e.g. changing content source, using REX).

ehelms commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -112,25 +112,50 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
-        yum localinstall $CONSUMER_RPM -y
+    if ! command -v subscription-manager &> /dev/null
+    then
+      yum install -y subscription-manager
```
I'd consider updating this to be safe for use with both `dnf` and `yum` for when the day comes they remove `yum`.

ares replied:
also one of the common request is to make sure we have the latest subscription-manager version available, could we address it here at the same time?

ehelms commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -112,25 +112,50 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
+    fi
+
+    # Prepare SSL certificate
+    cp -f $SSL_CA_CERT $KATELLO_SERVER_CA_CERT
```
I see that we copy this certificate from a temporary file created to hold it on the system as per (https://github.com/theforeman/foreman/pull/8563/files#diff-38bb9975863fa29a54da117e07e9bd666c02c3438261acafe31318c0f6a86567R39-R42). Do we then clean up that temporary location?

ehelms commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -112,25 +112,50 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
 
-    CONSUMER_RPM=$(mktemp --suffix .rpm)
-    curl --silent --show-error --output $CONSUMER_RPM <%= subscription_manager_configuration_url(hostname: @url_host) %>
+    KATELLO_SERVER_CA_CERT=/etc/rhsm/ca/<%= @url.host %>.pem
```
Currently, this file is:

```
/etc/rhsm/ca/katello-server-ca.crt
```

I think that mostly only matters due to how many users are used to this convention. I also think that naming the CA certificate after the host could be confusing to users. I'd vote stick with the way it is today `katello-server-ca.crt` or let's aim for some new naming convention that we want to aim for across the ecosystem (e.g. foreman_ca.crt, <hostname>-ca.crt)

ehelms commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -112,25 +112,50 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
+      --server.port="<%= @url.port %>" \
+      --server.prefix="/rhsm" \
+      --rhsm.repo_ca_cert="$KATELLO_SERVER_CA_CERT" \
+      --rhsm.baseurl="<%= @url %>pulp/content"
```
I think we still need to include https://github.com/theforeman/puppet-foreman_proxy_content/blob/master/lib/puppet/provider/rhsm_reconfigure_script/rhsm_reconfigure_script.rb#L119-L124

ehelms commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -112,25 +112,50 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
+      update-ca-trust enable
+      cp -f $KATELLO_SERVER_CA_CERT /etc/pki/ca-trust/source/anchors
+      update-ca-trust
+    fi
```
In talking with @jlsherrill -- this was originally added to support docker and I think we can drop this going forward. There has been at least one issue I've seen asking not to do this unless its needed (https://bugzilla.redhat.com/show_bug.cgi?id=1927020). 

If there is a service that users need to run on a client to talk to Foreman or Katello or Pulp, I think we can guide them to configure said service appropriately rather than trusting the typically self-signed certificates system wide.

ehelms commented:
@jlsherrill @jturel @evgeni This touches on areas you all have history and familiarity with, and I am raising questions about how we can tweak this to have a clean implementation moving forward and leave the baggage back in https://github.com/theforeman/puppet-foreman_proxy_content/blob/master/lib/puppet/provider/rhsm_reconfigure_script/rhsm_reconfigure_script.rb#L32. I would appreciate if y'all could each review thoroughly, making sure to look back at the current script to ensure we do not have any regressions but that we also do not carry forward any cruft if it can be avoided.

evgeni commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -112,25 +112,62 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
-        rpm -ivh --nodigest --nofiledigest $CONSUMER_RPM
+    # Install or upgrade subscription-manager
+    if ! [ -x "$(command -v subscription-manager)" ] ; then
+      if [ -f /usr/bin/dnf ]; then
+        dnf install -y subscription-manager
+      else
+        yum install -y subscription-manager
+      fi
```
In theory, we do also have sub-man on Debian (but not supported today by `bootrstrap.py`, and this line is guarded by "if RHEL or CentOS").

And I always think that the sheer presence of a binary should not mean it can be used blindly. My Debian systems have `/usr/bin/rpm` and my Fedora has `/usr/bin/dpkg` (those are not the binaries you're testing for, but you get the idea).

I think the correct approach should be to parse `/etc/os-release` and only if that fails fall back to "is that binary there". Probably good as a helper somewhere, as I expect this script to need "the right package manager" in a few places.

evgeni commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -112,25 +112,62 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
+
+    if ! [ -f $RHSM_CFG ] ; then
+      echo "'$RHSM_CFG' not found, cannot configure subscription-manager"
+      exit 1
```
This `exit` means the cleanup below never triggers? Or am I reading the flow wrongly?

ehelms replied:
I think you are reading this correctly.

evgeni commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -112,25 +112,62 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
 
-    CONSUMER_RPM=$(mktemp --suffix .rpm)
-    curl --silent --show-error --output $CONSUMER_RPM <%= subscription_manager_configuration_url(hostname: @url_host) %>
+    KATELLO_SERVER_CA_CERT=/etc/rhsm/ca/katello-server-ca.crt
```
The old path was `/etc/rhsm/ca/katello-server-ca.pem`

evgeni commented:
One thing that I realized while looking at this, but that is not explicitly related.

There are many scripts out there, which check "is this machine subbed to katello/satellite" and if we change anything in the local representation (on the client machine), those might fail. I know we want to get rid of the RPM, and that's fine, but we need to very loudly communicate this and also make sure all other "markers" (like the `/etc/rhsm/ca/katello-server-ca.pem` path) should remain the same.

Example: https://github.com/Katello/katello-client-bootstrap/blob/e6ce46b62fbb3414a2065facf5cb420801ac3fe5/bootstrap.py#L355-L360

ehelms commented:
> There are many scripts out there, which check "is this machine subbed to katello/satellite" and if we change anything in the local representation (on the client machine), those might fail. I know we want to get rid of the RPM, and that's fine, but we need to very loudly communicate this and also make sure all other "markers" (like the `/etc/rhsm/ca/katello-server-ca.pem` path) should remain the same.
> 
> Example: https://github.com/Katello/katello-client-bootstrap/blob/e6ce46b62fbb3414a2065facf5cb420801ac3fe5/bootstrap.py#L355-L360

These paths or at least names, will change in the future. I think then we should aim for some more reliable, and "API"-like way of indicating that a system is registered to a Foreman. Either laying down a file somewhere on the machine, or working to get a config value in sub-man/rhc that indicates this going forward. I'll happily go write up an issue somewhere if we can agree to what we think is the best approach to that. I think today if you were using pure Global Registration without Katello or Puppet, you would have no indicator that the system was tied to a Foreman instance right? So would a file laid down somewhere that is more generic make sense for broader use cases?

evgeni commented:
> These paths or at least names, will change in the future.

I think we should then try to teach at least bootstrap about the new paths and/or the API-like thing you mention, as even if it's deprecated, I see users using it in the future, and it should properly detect the system being subscribed.

And yes, today, without Kat/Pup, you have no way two know who manages the system.

ekohl commented:
This is going a bit off topic, but I'm not sure where else to discuss it.

In Puppet speak, I think we should come up with some fact that returns the host. From a clients perspective, you know the content and RHSM URLs for sure. Is that sufficient or do you want to know more?

From a configuration management perspective, what I'd really like is for the ENC to provide this information. Then you can use Puppet/REX/Ansible to correct the registration. This should allow users to move hosts between content proxies.

If you provide it in both the ENC and as facts, you can also detect out-of-sync configs and correct them.

Back to this PR: if you have that info in the ENC, you already have the helpers to provide it as a macro here.

ehelms commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -112,31 +117,70 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
+
+    subscription-manager register <%= '--force' if @force %> \
+      --org='<%= @organization.label %>' \
+      --activationkey='<%= activation_keys %>' || <%= @ignore_subman_errors ? 'true' : 'cleanup_and_exit 1' %>
```
Sorry if I should know this, does this mean that I can only register with an activation key? What happens if I want to register with username and password?

stejskalleos replied:
Yes, registration with activation key is the only workflow supported right now. If users wants to use login & password they have to edit the template and add parameters manually

ehelms replied:
Is username and password support on the Roadmap?

evgeni replied:
fwiw, bootstrap doesn't support username/pass either today.

ehelms replied:
Bootstrap does so many things and that is one it does not do? TIL. Let me try to just write out the scenario that I am envisioning and ask a question at the end for discussion of how we want things to work for users.

The move to have GR directly re-configure subscription-manager allows us to drop the bootstrap RPM and centralize on the GR API as **the** way to register. Today, users can and are used to installing the bootstrap RPM and then calling subscription-manager to register a host with username and password or activation key. This current design enforces using an activation key. 

Is that a design choice we want to make? And if yes, how would it affect users who are used to username and password?

evgeni replied:
But today users cannot, neither bootstrap nor the registration templates can call sub man with username/password.

So if users want this, they need manual work anyways. And I personally don't see a goid use for user/pass.

ehelms replied:
Just to make sure I am clear, today users can:

 1) yum install katello-ca-consumer-latest RPM
 2) subscription-manager register --username --password --org

You are saying that we should drop both parts of this workflow and users should always create at least one activation key to do any registration of content hosts?

ares replied:
The manual editting of the initial curl still accepts username and password though, right? but it's a valid question, do we need to support username and password and where one would specify that in the reg form? Would that be considered secure if the password is passed in the plain text form? I know that AK name is basically the same, but it's not called "password" and can be used more in a form of a token :-) Plus it's authorization scope is limited.

ekohl replied:
I think that in our template we shouldn't really care about username/password. AKs do have benefits and I can imagine that at some point we actually generally a new key on the fly and automatically expire it after it's been used when this workflow is used. Like a one-time-activation-key. Not something to implement here, but having a registration workflow inside Foreman that supports it does make that flow possible.

ehelms replied:
I know this is "registration", but throwing the idea out there anyway. If a user does not supply an activation key we could opt to configure subscription-manager but not register with subscription-manager and output a message saying this must be done manually by calling subscription-manager. That would allow for this existing workflow to work.

I am OK if we drop username/password support. I would want us to be clear about that and call it out as a deprecated or removed (depending on timeline) feature for users in our release notes.

I'd like to also get @jlsherrill and @jturel to think on this.

jlsherrill replied:
I am too kinda hesitant to completely remove this functionality.  Its generally the scenario that developers use, and i wouldnt' be surprised if users did too.  Its even mentioned in the docs: https://docs.theforeman.org/nightly/Content_Management_Guide/index-katello.html

I like the workflow you mention, if an activation key isn't present, print out the sub-man command without the activation key and let the user run it, alternatively, it could run the command and let them input their user/pass directly?

ekohl replied:
I don't think it should print a command to run: the registration should either work (and be complete) or fail for a technical reason. It shouldn't start with a TODO. I'm also hesitant to make the script interactive.

Given bootstrap doesn't support it either, are we really dropping anything? It feels more like a possible future RFE.

evgeni replied:
Yeah, the above command has feature-parity with bootstrap, everything else is RFE and/or "not the main workflow we support, you can script it yourself"

ares commented:
> Back to this PR: if you have that info in the ENC, you already have the helpers to provide it as a macro here.

@ekohl while I like the idea of adding an info provider that would define the host rhsm source, this would only be useful for existing hosts. We're talking about registration template here, the host at that point does not exist. We'd only have this available in host-init-config template (the second step) but that's after we configured the rhsm already. It would be great for ongoing configuration of a registered machine.

ekohl commented:
> @ekohl while I like the idea of adding an info provider that would define the host rhsm source, this would only be useful for existing hosts. We're talking about registration template here, the host at that point does not exist. We'd only have this available in host-init-config template (the second step) but that's after we configured the rhsm already. It would be great for ongoing configuration of a registered machine.

Correct, but my point is that I want some function that returns the correct value and use that function wherever you need the URL. That could be an ENC or a template macro. That gives us a single place of truth.

Hardcoding it in the template is IMHO always wrong. Perhaps should be provided via the context with a TODO to get it directly from Katello. That way we're at least prepared for the future and template authors know not to rely on hardcoded paths.

ehelms commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -112,31 +117,70 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
+    test -f $RHSM_CFG.bak || cp $RHSM_CFG $RHSM_CFG.bak
+    subscription-manager config \
+      --server.hostname="<%= @url.host %>" \
+      --server.port="<%= @url.port %>" \
```
What does this port value resolve to? The reason I ask is that today, if a host is registering to Foreman this would be 443, if a host is registering to a content proxy this port would be 8443. We have a story to change this on a content proxy to be 443 and thus align it with the server. That change would make this registration story cleaner over all but we would need to land that change first. We are starting work on it very soon.

ekohl replied:
Btw, this is the reason why I really am insisting that we use helpers for this. Maybe I should have made it even more explicit.

stejskalleos replied:
Hmm, TBH totally forgot about smart-proxy! So what should be the behavior here? Do we want to `subscription-manager` call the proxy or ignore the proxy's URL & set `server.hostname` & `server.port` to Foreman?

evgeni replied:
The proxy! Right now port 8443 for RHSM and port 443 for pulp, to be consolidated later into 443.

ekohl replied:
I decided to spend a bit of time to come up with a partial patch. https://github.com/Katello/katello/pull/9413 has the following code:
```ruby
def rhsm_url
  # TODO: We shouldn't relate this to the Pulp feature and also expose it
  # as a setting
  if pulp_mirror?
    "https://#{URI.parse(url).host}:8443/rhsm"
  elsif pulp_primary?
    "https://#{URI.parse(url).host}/rhsm"
  end
end
```

So you can convert this to a host macro:
```ruby
def host_rhsm_url(host)
  return unless host.respond_to?(:content_source) && host.content_source

  smart_proxy = host.content_source
  # TODO: We shouldn't relate this to the Pulp feature and also expose it
  # as a setting
  if smart_proxy.pulp_mirror?
    "https://#{URI.parse(smart_proxy.url).host}:8443/rhsm"
  elsif smart_proxy.pulp_primary?
    "https://#{URI.parse(smart_proxy.url).host}/rhsm"
  end
end
```

stejskalleos replied:
Hmm, not sure how to do it in the `global_registration` template.
* I cannot use `host_rhsm_url(host)` because I don't have `host` object yet (host is not created yet).
* The only way to get URL is from `url` parameter or take Foreman's default url [0]
* And I can't use `pulp_mirror? || pulp_primary?` because I don't know if the URL passed to `/register?url=...` [0] is smart-proxy's  URL or something else.

[0] https://github.com/theforeman/foreman/blob/develop/app/controllers/concerns/foreman/controller/registration.rb#L85-L96

ekohl replied:
You do have a Smart Proxy instance right? That's your content source instance. If not, you should use SmartProxy::pulp3_primary I believe.

Note that to be really safe, you should verify the Pulp 3 feature is present on that Smart Proxy.

stejskalleos commented:
Rebased & updated,
fixed issue with copying katello-ca cert & added URL helper for `rhsm_base_url`, see https://github.com/Katello/katello/pull/9417

ehelms commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -112,31 +117,70 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
 
-    CONSUMER_RPM=$(mktemp --suffix .rpm)
-    curl --silent --show-error --output $CONSUMER_RPM <%= subscription_manager_configuration_url(hostname: @url_host) %>
+    # Prepare subscription-manager
```
Maybe something like:

```suggestion
    # katello-ca-consumer* is the old way to configure subscription-manager replaced by Global Registration
    # Clean up any previous katello-ca-consumer* RPMs before re-configuring subscription-manager
```

ekohl commented on app/controllers/concerns/foreman/controller/registration.rb:
```diff
@@ -83,18 +82,18 @@ def not_found(options = nil)
   end
 
+  def context_urls
+    { server_url: URI(root_url), registration_url: registration_url }
```
server is a highly ambiguous word in the Foreman context. Can we be more explicit which server? I also think it's no longer used now?

ekohl commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -112,31 +117,70 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
+    subscription-manager config \
+      --server.hostname="<%= @rhsm_url.host %>" \
+      --server.port="<%= @rhsm_url.port %>" \
+      --server.prefix="/rhsm" \
```
Can't we use the path from the URL? Even if today it doesn't change, it keeps all logic for the RHSM URL in a single place. It also makes it clear that this is all from the same value
```suggestion
      --server.prefix="<%= @rhsm_url.path %>" \
```

ekohl commented on app/controllers/concerns/foreman/controller/registration.rb:
```diff
@@ -83,18 +82,18 @@ def not_found(options = nil)
-          else
-            URI(register_url)
-          end
+    uri = URI(params[:url] || root_url)
```
Is it explicit that this no longer uses `register_url`?

ekohl commented on app/controllers/concerns/foreman/controller/registration.rb:
```diff
@@ -82,19 +81,19 @@ def not_found(options = nil)
   end
 
+  def context_urls
+    { url: url, registration_url: URI.join(url, 'register') }
```
I think `url` is still ambiguous and also something we previously didn't have. Why is it needed?

ekohl commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -112,31 +117,70 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
+    # Configure subscription-manager
+    test -f $RHSM_CFG.bak || cp $RHSM_CFG $RHSM_CFG.bak
+    subscription-manager config \
+      --server.hostname="<%= @url.host %>" \
```
This still doesn't use the `rhsm_url` helper.

ekohl commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -126,25 +130,65 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ] || [ x$ID = xol ]; then
-    if [ "$(cat /proc/sys/crypto/fips_enabled)" = "1" ]; then
-        rpm -ivh --nodigest --nofiledigest $CONSUMER_RPM
+    if ! [ -x "$(command -v subscription-manager)" ] ; then
+      if [ "${VERSION_ID:0:1}" -gt 7 ]; then
```
Why do you need to get a substring of the version here? It doesn't work no Fedora since `3 < 7` and will break once we get to RHEL 10.

Or is it to deal with `7.7`? If so, then I'd prefer to strip `.*` (note it's not a real regex, . is a literal value):
```suggestion
      if [ "${VERSION_ID%.*}" -gt 7 ]; then
```

ekohl reviewed (dismissed):
I am good with this change. Given other reviewers had a comments, I'll give them some time to react.

ehelms commented:
I am attempting to test a few of the nuances that come with this change and am realizing I am missing something. I am trying the following but I get unauthorized errors:

```
curl -sS "https://$HOSTNAME/register?activation_keys=Test+AK&organization=Test_Organization"
```

ekohl commented:
AFAIK you need to provide a username/password to access the registration. You can also use the UI to get a token. That should generate a command for you.

ehelms commented:
> AFAIK you need to provide a username/password to access the registration. You can also use the UI to get a token. That should generate a command for you.

If that is the case, then we have a gap in functionality. Today, I can use an activation key to register via subscription manager without the need for any other auth mechanism.

ares commented:
> If that is the case, then we have a gap in functionality. Today, I can use an activation key to register via subscription manager without the need for any other auth mechanism.

Didn't we agree on rather providing a safer option - meaning a generated token? For people who would still want to construct the curl manually and run it from the client, they could use username/password with personal access token instead of the password. Since users who want to use this IMHO better registration method will have to learn new command (sub-man vs curl), can they also learn to use another kind of token?

Also FTR, today you can still use subscription-manager with AK directly. However for all the additional logic the registration or host-init template may contain, I think it's better to properly identify a user and authneticate using some stronger secret.

EDIT: and one more comment, we wanted to long term move to the model where use uses the HG to define everything for the host. If we start by using AK for authentication, how it should then behave when people use it also for content access definition and would want to use more keys?

ehelms commented:
> Didn't we agree on rather providing a safer option - meaning a generated token? For people who would still want to construct the curl manually and run it from the client, they could use username/password with personal access token instead of the password. Since users who want to use this IMHO better registration method will have to learn new command (sub-man vs curl), can they also learn to use another kind of token?

I recall talking about this as the desired long term push for users. My brain is fuzzy on cutting off the currently supported workflow for activation keys. That sort of leaves us with two choices I think:

 * Global registration, which with this change, can now directly update subscription-manager configuration enforces the use of an auth mechanism to register
 * Workflows that desire to use activation keys as they are allowed to be used now must continue to use the consumer RPM + subscription-manager directly

This would effectively mean we cannot drop the consumer RPM until we are ready to cut users off from using activation keys as their sole method of connecting a host. And I think that sort of change needs communication and a deprecation warning to enable making the larger change to GR only in a future release. As while a user could manually configure subscription-manager, it is a widely known and accepted workflow to do consumer RPM + activation key.

Tell it to me like I'm 5 years old, what are the steps I need to take to support the following workflow? (I'm working on ensuring we have a few GR tests in bats and can't use the UI).

I run a Foreman+Katello, I sync and publish content for consumption by anonymous users. Today, I generate an activation key and publish to users that they can download the consumer RPM, and register with said activation key. Users are then able to connect, and receive the content I published for their own testing.

ares commented:
Well, we don't have to drop the katello-ca-consumer right away if we need more time for deprecation. However introducing something in the middle would be harder to replace later.

The use case you talk about has two parts. One is the Foreman+Katello admin setups the instance. Second part is user then registers and consumes from other machine. That user must know the following - the Foreman url and activation key name.

With this new registration, the first part is the same with few additional steps. Imagine this admin wants to allow anonymous consumption based on the AK knowledge. She first creates the anonymous user with just enough permissions required. Then she uses impersonation feature to become the anonymous user. And then navigates to the register host screen. In advanced tab she checks the "unlimited" token life time and generates the command. That results into something like

```
curl -sS  'https://foreman.example.com/register?activation_keys=rhel8&organization_id=3' -H 'Authorization: Bearer eyJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjo0LCJpYXQiOjE2MjY3OTI0MTEsImp0aSI6IjFmYWVjNTdiOTA2MmZjMjJiNzZkNTViMjFmZTlhOTIzMzZhZGJhZTBlOTkzM2U0ZjZlM2YzYzhkMmZiNWEzMzgiLCJzY29wZSI6InJlZ2lzdHJhdGlvbiNnbG9iYWwgcmVnaXN0cmF0aW9uI2hvc3QifQ.jjvPEII9zsAxJ-pP8XVg1s9pMINMF2rYqU-ciwRIW6I' | bash
```

She will share the command and explain everyone who uses it that they can modify the rhel8 AK in the url with whatever they want to consume from the list of AKs she will also share.

If in part 2 users complain that the Authorization token is too long to type manually (haha), she will adjust the command with the credentials used in the first step.

```
curl -sS  'https://anonymous:password@foreman.example.com/register?activation_keys=rhel8&organization_id=3' | bash
```

again with the instructions the user can modify the activation_keys param. If she doesn't like the fact users can now login to the UI, she can in fact create a personal access token instead. That only grants access to the API. The command would then look like this

```
curl -sS  'https://anonymous:pat_token@foreman.example.com/register?activation_keys=rhel8&organization_id=3' | bash
```

she could even change her mind later and expire the personal access token, while she does not have to touch the existing AK names.

If you think that's overly complicated, some time ago we talked about some shell wrapper around curl command. Perhaps yggdrasil could be used to make this slightly easier.

Also to mention, the admin does not have to use UI for any of the setup operations. There is a `POST /api/registration_commands` to obtain the curl command as well. Users could generate their own commands too, but they need username and password/pat_token for that. Perhaps that would be the ultimate experience if we integrate with yggd. But so far, the curl command was acceptable user interface.

ares commented:
@stejskalleos this needs another rebase ;-)

ehelms commented:
Thanks for that detail @ares . I would agree that "special" user account + token would cover the workflow with just an activation key and as discussed that we should document that workflow for users as the replacement or as the way to grant users access to content without wanting to have a user account or manage the host through Foreman. That will give us the run way to move away from consumer RPM and promote the existing ways users register through content workflows but with global registration.

A note of at least two spots worthy of updates:

 * https://github.com/theforeman/foreman-documentation/blob/b74f8d5da245efb77f7106cb8fc13043ca76949e/guides/doc-Content_Management_Guide/topics/Managing_Activation_Keys.adoc#using-activation-keys-for-host-registration
 * https://github.com/theforeman/foreman-documentation/blob/b74f8d5da245efb77f7106cb8fc13043ca76949e/guides/doc-Content_Management_Guide/topics/QuickStart.adoc#registering-a-centos-stream-host-to-project


Additionally, based on the above I am going to look at adding some bats tests to verify the workflows.

ares commented:
Also ack from @evgeni @ehelms @jlsherrill would be great prior merging, I'm under the impression that all is resolved. Still needs to be tested

evgeni commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -42,6 +42,11 @@ cat << EOF > $SSL_CA_CERT
 EOF
 
+cleanup_and_exit() {
+  rm -f $SSL_CA_CERT
```
this doesn't cleanup `KATELLO_SERVER_CA_CERT` or roll-back the changes to `rhsm.conf`

is that intentional?

evgeni replied:
(probably, because it's *also* called in the success case, but I think we should be nicer in the error case)

ehelms replied:
I agree with this. Today, on uninstall users set back to `rhsm.conf.bak`. If this fails, it would be nice to restore them to whatever they had originally so they are not left in a broken state. That is, on failure restore the rhsm.conf.bak to rhsm.conf. Use cases:

1) User is moving a host from being registered to Red Hat CDN to a Foreman, something fails, they would get reverted and remain connected and operational
2) User is moving a host from one Foreman to another Foreman, this would prevent the host from being in limbo if it failed

evgeni replied:
Two Foremans won't work, as the first registration has created a .bak file, so the second registration won't create one (and then restore the original rhsm.conf, from before the first registration, on failure).

But this is probably a corner case we can fix later

stejskalleos replied:
> this doesn't cleanup KATELLO_SERVER_CA_CERT or roll-back the changes to rhsm.conf
> 
> is that intentional?

But do we want to remove `KATELLO_SERVER_CA_CERT`? If we remove the cert  then all communication between `subscription-manager` & foreman will fail with SSL error, right?

> Today, on uninstall users set back to rhsm.conf.bak

We should do the backup of the `rhsm.conf` that's for sure.

> If this fails, it would be nice to restore them to whatever they had originally so they are not left in a broken state. That is, on failure restore the rhsm.conf.bak to rhsm.conf.

Is this something what we do right now with `katello_ca_consumer.rpm`? I mean, if registration with `rpm` fails, does it do all the rollback steps automatically or users have to do them manually?

Because implementing rollback feature (while catching all the possible fail scenarios) won't be easy.
We need to cleanup methods:
* When script is successful
* When script failed

IMHO this is out of the scope of this PR, I would suggest to do it in separate PR and not block this one

ehelms replied:
> But do we want to remove `KATELLO_SERVER_CA_CERT`? If we remove the cert then all communication between `subscription-manager` & foreman will fail with SSL error, right?

This and the rhsm.conf roll back is tricky, as today we do this if the RPM is removed (https://github.com/theforeman/puppet-foreman_proxy_content/blob/master/lib/puppet/provider/bootstrap_rpm/bootstrap_rpm.rb#L203). But I do not see us doing it if the re-configure script fails. I would agree that could be left for a follow up given this does reflect current state.

evgeni commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -129,25 +133,65 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ] || [ x$ID = xol ]; then
-        rpm -ivh --nodigest --nofiledigest $CONSUMER_RPM
+    if ! [ -x "$(command -v subscription-manager)" ] ; then
+      if [ "${VERSION_ID%.*}" -gt 7 ]; then
+        dnf install -y subscription-manager
```
I might have said that before, but it feels overly complicated to have this if/else for version > 7 here, but not using the same in L141 where you remove ca-consumer, or L132/133 where you poke around sub-man.

given `yum` is an alias for `dnf` on EL8, I'd probably just use `yum` everywhere for now, and later introduce a central function that does the if/else?

stejskalleos replied:
> I'd probably just use yum everywhere for now, and later introduce a central function that does the if/else?

It was like that before but it wasn't accepted well by community so I had to change it to the current logic

ehelms replied:
I had suggested trying to future proof this more. I do get the point @evgeni is raising here, this logic should be applied in all cases of the use of yum or dnf and thus having a function to wrap the logic so it can be re-used everywhere is much cleaner.  Whether thats here now or in a follow up.

If we can get an issue that tracks making that update to determine yum vs. dnf using a single function I'm good with just using yum for now.

Neither option blocks for me. But the lack of consistency ought to be tracked and addressed.

evgeni replied:
Yeah, I'm all for future proofing this. But then *all* calls the $packagemanager need to be.

evgeni commented:
Left two comments, but workflow-wise this looks good to me!

ehelms commented:
I opened a PR to add some bats tests for global registration: https://github.com/theforeman/forklift/pull/1391
This is not a blocker to this PR but rather inspired by the PR that we should add tests across the workflows to ensure as we transition our expectations continue to work.

stejskalleos commented:
Rebased & Updated
* Now creating backup of `rhsm.conf`
* added method for choosing package manager
* created [redmine task](https://projects.theforeman.org/issues/33156) for cleanup & rollback methods

ehelms commented:
[test upgrade]

ehelms reviewed (dismissed):
I would like to get https://github.com/theforeman/forklift/pull/1391 resolved and available so that we have assurances that changes like this in the future don't break our expected workflows.

ares commented:
will require a rebase now

stejskalleos commented:
Rebased & updated (https://github.com/theforeman/foreman/commit/e2918f95f2c11bef3a287f63949387d47c771202)

ehelms commented:
[test katello]

stejskalleos commented:
Rebased & updated: fixed following errors when registering `CentOS 8`:

```
cp: cannot stat '/etc/rhsm/rhsm.conf': No such file or directory
cp: cannot create regular file '/etc/rhsm/ca/katello-server-ca.pem': No such file or directory
chmod: cannot access '/etc/rhsm/ca/katello-server-ca.pem': No such file or directory
```

ekohl commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -37,11 +37,36 @@ if [ -f /etc/os-release ] ; then
+# apt-get for Debian & Ubuntu
+# dnf for Fedora (version >= 22) & RHEL family (version > 7)
+# yum for Fedora (version < 22) & RHEL family (version < 8)
+if [ x$ID = xrhel ] || [ x$ID = xol ] || test "${ID_LIKE#*rhel*}" != "$ID_LIKE" ; then
```
Elsewhere we moved to this:
```suggestion
if [ -f /etc/redhat-release ] ; then
```

ekohl commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -37,11 +37,36 @@ if [ -f /etc/os-release ] ; then
+  else
+    PKG_MANAGER='yum'
+  fi
+elif [ x$ID = xdebian ] || [ x$ID = xubuntu ]; then
```
And this
```suggestion
elif [ -f /etc/debian_version ]; then
```

stejskalleos commented:
Rebased with changes from https://github.com/theforeman/foreman/pull/8789

stejskalleos commented:
[test katello]

stejskalleos commented:
@evgeni @ekohl @jlsherrill @ehelms 

PR updated, rebased & all green.
Can we merge? It's been here for a while & I have some other PR waiting for this one.

evgeni commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -37,11 +37,36 @@ if [ -f /etc/os-release ] ; then
+if [ x$ID = xfedora ]; then
+  if [ "${VERSION_ID%.*}" -gt 21 ]; then
+    PKG_MANAGER='dnf'
+  else
+    PKG_MANAGER='yum'
+  fi
```
I don't think we need to support Fedora <21

evgeni commented:
I think all my points were addressed, so :+1:

ares reviewed (approved):
Given two acks and the fact I've tested the functionality I think this can finally get it! @ehelms bats tests were also already merged, so I'm merging this.

Thanks @stejskalleos for a nice improvement, we can now officially deprecate the katello_ca_consumer rpm. Thanks @ehelms @ekohl and @evgeni for reviewing.

ezr-ondrej commented:
Is this worth for adding a Headline feature?

ekohl commented:
For Foreman itself I wonder if it's useful since it's heavily reliant on Katello. Katello should certainly mention this in their release notes.
